File: bindings/python/examples/device/capture_microphone.py
"""Microphone capture quick start — raw PCM.

✅ REAL — the microphone capture capability is implemented in the native C ABI
(`mediaway_ffi`, `mediaway_audio_capture_*`); this example runs against
it. Mirrors `examples/device/capture_microphone.rs`: open the default mic,
poll ~2 seconds of raw interleaved f32le PCM frames, print the negotiated
format, close. No encoding — there is no audio encoder in the ABI.
"""

import time

from mediaway import AudioCapture


def main() -> None:
    try:
        mic = AudioCapture.open(sample_rate=48000)
    except Exception as err:  # DeviceUnavailableError etc. — expected outcome
        print(f"no microphone on this machine: {err}")
        return

    with mic:
        rate = mic.sample_rate()
        channels = mic.channels()
        print(f"mic negotiated: {rate} Hz, {channels} ch")

        frames = 0
        total_bytes = 0
        deadline = time.monotonic() + 2.0
        while time.monotonic() < deadline:
            data = mic.poll_pcm()
            if data is not None:
                total_bytes += len(data)
                frames += 1
            else:
                time.sleep(0.002)

    print(f"captured {frames} PCM frame(s), {total_bytes} bytes in 2 s")


if __name__ == "__main__":
    main()

File: bindings/nodejs/examples/device/capture-microphone.ts
/**
 * capture-microphone.ts — microphone capture quick start (raw PCM).
 *
 * Status: ✅ real ABI under it.
 * The C ABI's microphone capture (raw interleaved f32le PCM,
 * mediaway_audio_capture_*) is implemented; this example runs against it via
 * @mediaway/device. Mirrors examples/device/capture_microphone.rs: open the
 * default mic, poll ~2 s of PCM frames, print the negotiated format, close.
 * No encoding — there is no audio encoder in the ABI.
 *
 * Run: npx tsx examples/device/capture-microphone.ts
 */

import { openMicrophone, CaptureUnavailableError } from "@mediaway/device";

const RECORD_MS = 2_000;
const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

async function main(): Promise<void> {
  let mic;
  try {
    mic = await openMicrophone({ sampleRate: 48_000, channels: 2 });
  } catch (err) {
    if (err instanceof CaptureUnavailableError) {
      console.log("no microphone on this machine; nothing to capture");
      return;
    }
    throw err;
  }

  try {
    let frames = 0;
    let totalBytes = 0;
    const startedAt = Date.now();
    while (Date.now() - startedAt < RECORD_MS) {
      const audio = mic.pollFrame();
      if (audio === null) {
        await sleep(2);
        continue;
      }
      if (frames === 0) {
        console.log(`mic negotiated: ${audio.sampleRate} Hz, ${audio.channels} ch`);
      }
      frames++;
      totalBytes += audio.data.length;
    }
    console.log(`captured ${frames} PCM frame(s), ${totalBytes} bytes in ~${RECORD_MS} ms`);
  } finally {
    await mic.close(); // joins the backend worker thread
  }
}

main().catch((err) => {
  console.error(err);
  process.exitCode = 1;
});
